streams/src/main/java/org/apache/kafka/streams/state/internals/InMemoryTimeOrderedKeyValueBuffer.java

```diff
@@ -259,23 +254,23 @@ public void flush() {
     }
 
     private void logValue(final Bytes key, final BufferKey bufferKey, final ContextualRecord value) {
-        final byte[] serializedContextualRecord = value.serialize();
+        final byte[] innerValue = value.value();
 
         final int sizeOfBufferTime = Long.BYTES;
-        final int sizeOfContextualRecord = serializedContextualRecord.length;
+        final int sizeOfContextualRecord = innerValue.length;
 
         final byte[] timeAndContextualRecord = ByteBuffer.wrap(new byte[sizeOfBufferTime + sizeOfContextualRecord])
                                                          .putLong(bufferKey.time)
-                                                         .put(serializedContextualRecord)
+                                                         .put(innerValue)
                                                          .array();
 
         collector.send(
             changelogTopic,
             key,
             timeAndContextualRecord,
-            V_1_CHANGELOG_HEADERS,
-            partition,
             null,
+            partition,
+            value.recordContext().timestamp(),
```

Comment:
Why do we add a timestamp now?

Reply:
Two reasons, maybe neither of them good. We used to set it this way, and it messes up the tests to let the producer set them. I can take another look at the restore to see if I missed something from the revert.

Reply:
What I am wondering is, why we removed it though via https://github.com/apache/kafka/commit/a193f370c0cd43d056dd913e97e457e636dbf76d#diff-6f5b9383b8acde07a74d681de2da66d2R256 (ie, we don't set the timestamp in `2.2`, `2.3`, `trunk` either).

Reply:
Actually, we do use it on restore, unless I've overlooked something.

IIRC, in the fixed versions (`2.2.1+`), we are serializing the entire record context, including the timestamp, in the value. There's no need to go the extra mile and also set the timestamp on the changelog record's timestamp field as well, so I just left it null and let the producer set it. I.e., the fix "moves" it into the value, not "copies" it.

Here, though, we do need to preserve the timestamp so we can at least restore that much information.

Reply:
So it's an additional fix for 2.1?

Reply:
No, I think this PR is all we need, unless I've missed something. We store it in the timestamp field and then extract it again on restore. We'll never be able to store it in the value in 2.1 the way we do in 2.2.1+, so this should be the "final result" :)